irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java

```diff
@@ -1645,15 +1663,101 @@ private static Inode toFh(long _inodeNumber)
+                    strPath = resultSet.getResults().get(0).getColumn(1); 
+                    log_.debug("path from coll_id [{}]", strPath);
+                }                                    
+            } catch (GenQueryBuilderException | JargonException | JargonQueryException e)
+            {
+                log_.error(e.getMessage());
+            }    
+        } else if (option.equals("data")) {
+            try {
+                builder = new IRODSGenQueryBuilder(true, null);
+    
+                builder.addSelectAsGenQueryValue(RodsGenQueryEnum.COL_D_DATA_ID)
+                    .addSelectAsGenQueryValue(RodsGenQueryEnum.COL_COLL_NAME)
+                    .addSelectAsGenQueryValue(RodsGenQueryEnum.COL_DATA_NAME)
+                    .addConditionAsGenQueryField(RodsGenQueryEnum.COL_D_DATA_ID, QueryConditionOperators.EQUAL, String.valueOf(_inodeNumber));
+                
+                // the constructed GenQuery is: SELECT DATA_ID, COLL_NAME, DATA_NAME WHERE DATA_ID = '<_inodeNumber>'
+    
+                gqe = factory_.getIRODSGenQueryExecutor(adminAcct_);
+                query = builder.exportIRODSQueryFromBuilder(50);
+                resultSet = gqe.executeIRODSQueryAndCloseResultInZone(query, 0, adminAcct_.getZone());
+    
+                log_.debug("found row is (data id) [{}]", resultSet.getResults());
+                if (resultSet.getResults().size() != 0) {
+                    // since DATA_NAME is only the filename (no path), we'll need to combine it with COLL_NAME to get the full path
+                    // column 1 = COLL_NAME, column 2 = DATA_NAME
+                    strPath = resultSet.getResults().get(0).getColumn(1) + '/' + resultSet.getResults().get(0).getColumn(2);   
+                    
+                    log_.debug("combined path is [{}]", strPath);
+                }                       
+            } catch (GenQueryBuilderException | JargonException | JargonQueryException e)
+            {
+                log_.error(e.getMessage());
+            }
+        } else {
+            log_.warn("option [{}] is not supported. choose from `data` or `coll`", option);
+        }
+
+        return strPath;
+    }
+
     private Path getPath(long _inodeNumber) throws IOException
     {
         Path path = inodeToPathMapper_.getPathByInodeNumber(_inodeNumber);
 
+        // if no path matches _inodeNumber, we'll need to find where we are from GenQuery
+        // since a path may lead to a data object or a collection, we'll need to check both
         if (path == null)
         {
-            throw new NoEntException("Path does not exist for [" + _inodeNumber + "]");
-        }
+            log_.debug("Path was null (maybe because the NFSRODS container was restarted without remounting), looking up in GenQuery");
+            String collPath = tryAndFindPathFromGenQuery(_inodeNumber, "coll");
+            String dataPath = tryAndFindPathFromGenQuery(_inodeNumber, "data");
```

Comment:
Seems these should be conditional.

```java
String path = tryAndFindPathFromGenQuery(_inodeNumber, "coll");

if (path.isEmpty())
{
    path = tryAndFindPathFromGenQuery(_inodeNumber, "data");

    if (path.isEmpty())
    {
        return null; // Or empty optional?
    }
}
```

Reply:
Yep, I agree. Instead of returning `null`, the original `getPath()` function threw a `NoEntException`,  so I'll keep that.

Reply:
Yep. The exception is important to NFSRODS working correctly.